### PR TITLE
Update S2DEX Text Engine to latest version; fix build issues

### PR DESCRIPTION
**Pull request by someone2639**
Now its just part of HackerSM64 entirely, and HackerSM64 will be the source of all changes (as long as said changes can reasonably apply to homebrew games)

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -124,6 +124,12 @@ endif
 TEXT_ENGINE := none
 $(eval $(call validate-option,TEXT_ENGINE,none s2dex_text_engine))
 
+ifeq ($(TEXT_ENGINE), s2dex_text_engine)
+  DEFINES += S2DEX_GBI_2=1 S2DEX_TEXT_ENGINE=1
+  SRC_DIRS += src/s2d_engine
+endif
+# add more text engines here
+
```
Comment (arthurtilly) [inline Makefile:132]: why was this moved

File: src/s2d_engine/Makefile [hunk 2]
```diff
@@ -10,6 +10,24 @@ DUMMY != mkdir -p $(BUILD_DIR)
...
+else ifneq ($(call find-command,mips-linux-gnu-ld),)
+  CROSS := mips-linux-gnu-
+else ifneq ($(call find-command,mips64-linux-gnu-ld),)
+  CROSS := mips64-linux-gnu-
+else ifneq ($(call find-command,mips-ld),)
+  CROSS := mips-
+else
+  $(error Unable to detect a suitable MIPS toolchain installed)
+endif
+
```
Comment (arthurtilly) [inline src/s2d_engine/Makefile]: whats point of this

File: src/s2d_engine/Makefile [hunk 3]
```diff
@@ -18,10 +36,9 @@ I_DIRS = src include/n64/PR include/n64 include/libc include .
 FULL_I_DIRS = $(addprefix $(INCLUDE_BASE),$(I_DIRS))
 I_FLAGS = $(foreach i,$(FULL_I_DIRS),-I$(i))
 
-TARGET_CFLAGS = -nostdinc -I include/libc -DTARGET_N64 -DF3DEX_GBI_2
+TARGET_CFLAGS = -nostdinc -I include/libc -DTARGET_N64 -DF3DEX_GBI_2 -DNON_MATCHING -DAVOID_UB
```
Comment (arthurtilly) [inline src/s2d_engine/Makefile]: huh?

File: src/s2d_engine/s2d_parse.c [hunk 4]
```diff
@@ -273,14 +273,15 @@ void s2d_type_print_alloc(int x, int y, int align, const char *str, int *pos) {
...
 
-// broken atm
 static int s2d_width(const char *str, int line, int len) {
 	char *p = str;
 	int tmp_len = 0;
 	int curLine = 0;
 	int width = 0;
 	int scale = 1;
 
+	u8 *ktable = seg2virt(s2d_kerning_table);
```
Comment (arthurtilly) [inline src/s2d_engine/s2d_parse.c:283]: what is this function/macro thing its weird

File: Makefile (hunk 1 above)
Comment (someone2639) [reply Makefile:132]: `SRC_DIRS` won't populate correctly unless this happens before `C_FILES` is set

File: src/s2d_engine/Makefile (hunk 2 above)
Comment (someone2639) [reply src/s2d_engine/Makefile]: whole file killed

File: src/s2d_engine/init.h [hunk 5]
```diff
@@ -5,8 +5,26 @@
...
+#elif   F3DEX2PL_GBI
+    #define zex_text gspF3DEX2_PosLight_fifoTextStart
+    #define zex_data gspF3DEX2_PosLight_fifoDataStart
+#elif   F3DEX_GBI_2
+    #define zex_text gspF3DEX2_fifoTextStart
+    #define zex_data gspF3DEX2_fifoDataStart
+#else
+    #error S2DEX Text Engine only works with other EX2 series microcodes
+#endif
+
```
[the remark this answers is not in the document]
Comment (someone2639) [reply src/s2d_engine/init.h:27]: Yes but only if you use F3DEX1 (only EX2 family microcodes can load between each other, or only EX1 microcodes)

File: src/s2d_engine/s2d_parse.c (hunk 4 above)
Comment (someone2639) [reply src/s2d_engine/s2d_parse.c:283]: homebrew compatibility where you can plug your own function into config.h

File: src/s2d_engine/init.h (hunk 5 above)
[the remark this answers is not in the document]
Comment (arthurtilly) [reply src/s2d_engine/init.h:27]: more so meant the massive table of ifdefs above this not the error itself

[the remark this answers is not in the document]
Comment (arthurtilly) [reply src/s2d_engine/init.h:27]: fair

**Review by arthurtilly — APPROVED** [review]
approved (but i hate you for the ifunny font)

File: src/s2d_engine/s2d_parse.c (hunk 4 above)
Comment (arthurtilly) [reply src/s2d_engine/s2d_parse.c:283]: coming back to this wouldnt it be better to have the function name still be segmented_to_virtual (same for all the config macros to have the default name the same as the sm64 name) so its less confusing and is still configurable for homebrew

**Review by gheskett — APPROVED** [review]
Just gonna assume this all works as intended™

This is opt-in after all so it's low risk to 2.1. At least lower risk than Puppyprint since that's used for debugging too.

**Review by gheskett — CHANGES_REQUESTED** [review]
On second thought...

image

Please fix these thank you :)

**Comment by someone2639** [comment]
warnings fixed

**Comment by gheskett** [comment]
Before I approve this, can we replace all uses of `IS_RUNNING_ON_EMULATOR` with gIsConsole for consistency sake?

File: src/s2d_engine/s2d_config.h [hunk 6]
```diff
@@ -88,7 +88,7 @@ extern char *proutSprintf(char *dst, const char *src, size_t count);
 #define TEX_BITDEPTH 8
 
 #define seg2virt segmented_to_virtual
-#define IS_RUNNING_ON_EMULATOR (IO_READ(DPC_PIPEBUSY_REG) == 0)
+#define gIsEmulator (!gIsConsole)
```
Comment (gheskett) [inline src/s2d_engine/s2d_config.h:91]: Well that's one way, not the biggest biggest fan but will approve nonetheless

**Review by gheskett — APPROVED** [review]
